parsl/dataflow/dflow.py

```diff
@@ -1277,6 +1277,19 @@ def cleanup(self) -> None:
             executor.shutdown()
             logger.info(f"Shut down executor {executor.label}")
 
+            if hasattr(executor, 'provider'):
+                if hasattr(executor.provider, 'script_dir'):
+                    logger.info(f"Closing channel(s) for {executor.label}")
+
+                    if hasattr(executor.provider, 'channels'):
+                        for channel in executor.provider.channels:
+                            channel.close()
+                    else:
+                        assert hasattr(executor.provider, 'channel'), "If provider has no .channels, it must have .channel"
+                        executor.provider.channel.close()
```

Comment:
For debugging / retrospective, would `log.debug()` for each closed channel be of any use?  (And if so, do we have recognizable output from `channel.__repr__` or `.__str__` to make that feasible?)

Reply:
yes to both: logging around closing is pretty useful here, even at INFO level, due to Parsl's history of hanging at shutdown; Channels are rendered as a recursive part of the start log-the-`Config`-object so they "should" already have an appropriate repr().